src/data/program_configuration.js

```diff
@@ -247,53 +278,41 @@ class CompositeExpressionBinder<T> implements Binder<T> {
 class ProgramConfiguration {
     binders: { [string]: Binder<any> };
     cacheKey: string;
-    layoutAttributes: ?Array<LayoutAttribute>;
-    PaintVertexArray: Class<StructArray>;
+    layoutAttributes: Array<StructArrayMember>;
 
-    paintVertexArray: StructArray;
-    paintVertexBuffer: ?VertexBuffer;
+    _buffers: Array<VertexBuffer>;
 
     constructor() {
         this.binders = {};
         this.cacheKey = '';
+
+        this._buffers = [];
     }
 
-    static createDynamic<Layer: TypedStyleLayer>(programInterface: ProgramInterface, layer: Layer, zoom: number) {
+    static createDynamic<Layer: TypedStyleLayer>(layer: Layer, zoom: number, filterProperties: (string) => boolean) {
         const self = new ProgramConfiguration();
-        const attributes = [];
-
-        for (const attribute of programInterface.paintAttributes || []) {
-            const property = attribute.property;
-            const name = attribute.name || property.replace(`${layer.type}-`, '').replace(/-/g, '_');
-            const value: PossiblyEvaluatedPropertyValue<any> = layer.paint.get(property);
+        for (const property in layer.paint._values) {
+            if (!filterProperties(property)) continue;
```

Comment:
The combination of `filterProperties` plus needing `instanceof` and `property-function` conditionals seems a bit cumbersome. Could we drive this loop via a passed in array of property names or something instead?